rfcs/2023-02-08-7496-vrl-return.md

```diff
@@ -0,0 +1,72 @@
+# RFC 7496 - 2023-02-02 - VRL Return Keyword
+
+Add return expression to the Vector Remap Language.
+
+## Context
+
+- #7496
+
+## Cross cutting concerns
+
+- None.
+
+## Scope
+
+### In scope
+
+- Adding a return expression to VRL.
+
+### Out of scope
+
+- Adding new keywords for similar purposes such as `drop`.
+- Defining semantics of keywords that are usually used for other purposes in other languages such as `break`.
+- Defining parameterized return, e.g. `return parse_json!(.message)`. The proposed return expression always passes current `.` to output and takes no input.
+
+## Pain
+
+- Aborting with changing the input cannot be easily done.
+- VRL code is often unnecessarily indented because of lack of early returns.
+
+## Proposal
+
+### User Experience
+
+- A `return` expression causes the VRL program to terminate, keeping any modifications made to the event.
+- The keyword is used by itself with no parameters and `.` will be sent to the output.
+
+### Implementation
+
+- Implementation will be largely the same as the current `abort` keyword when `drop_on_abort` is set to `false` and without the optional argument. The only difference is that the returned value will be taken from `.` and not from original input.
+- `drop_on_abort` will have no effect on return calls and configuration such as `drop_on_return` will not be added.
+
+## Rationale
+
+- It will be possible to write VRL with less indentation making it more readable.
+- `return` is already a reserved word so it can be used without introducing a breaking change.
+
+## Drawbacks
+
+- The `return` keyword will be given a semantic meaning that will have to be supported going forward.
+
+## Prior Art
+
+- Most languages have a way to make early returns.
+- There was no prior attempted implementation of returns in VRL to my knowledge.
+
+## Alternatives
+
+- New keywords that are not currently a reserved keyword can be added to the language. This would, however, constitute a breaking change.
+- This feature can also be rejected as it does not add any functionality that cannot be currently expressed.
+
+## Outstanding Questions
```

Comment:
What should happen if the `return` keyword is used in a closure like:

```
tally = {}
for_each(array!(.tags)) -> |index, value| {
    if index == 1 {
    return
  }
}
```

? My instinct is to disallow the `return` keyword inside of closures.

Reply:
A few options:
1. Disallow `return` inside closures
2. Stop the closure execution
3. Stop the VRL program and return

Reply:
I think I like option 1 the best. We could always migrate to option 2 or 3 in the future.

Reply:
I would expect option 2 because that's what most languages do. While 3 could be useful, I think that would be too surprising for many users so I would not use that with the `return` keyword.

For option 2, it also does not make sense to allow argumentless returns, because there's nothing like `.`, so I would deny them in the first iteration, if that's what we end up on.

Reply:
Agreed that option 2 is probably the most familiar to users. I was thinking of starting with option 1 purely because it'd likely be easier to implement, but if we can implement option 2 that would be preference.